pennylane/debugging.py

```diff
@@ -81,6 +82,9 @@ def circuit():
     """
 
     def get_snapshots(*args, **kwargs):
+        if qnode.interface == "auto":
+            qnode.interface = qml.math.get_interface(*args, *list(kwargs.values()))
```

Comment:
> Maybe use a with statement?

We could add a `set_interface` contextmanager method to the `QNode`:
```python
@contextlib.contextmanager
def set_interface(self, *args, **kwargs):
    if self.interface != "auto":
        yield
        return
    self.interface = qml.math.get_interface(*args, *list(kwargs.values()))
    yield
    self.interface = "auto"
    return
```
Such that we can do:
```python
with qnode.set_interface(*args, **kwargs):
    # do sth
```

Reply:
Here's another potential solution to this problem:

I believe this interface problem comes when wrapping `QNodes`, right? Thus all wrappers will have the same signature: `def wrapper(*args, **kwargs)`. We could then define the following method inside `QNode`:

```python
def set_interface(self, f):
    old_interface = self.interface
    def inner(*args, **kwargs):
      if self.interface == "auto":
          self.interface = qml.math.get_interface(*args, *list(kwargs.values()))
      result = f(*args, **kwargs)
      self.interface = old_interface
      return result
  return inner
```
Such that we can do:
```python
def transform(tape):
    def wrapper(*args, **kwargs):
        # do sth
    return tape.set_interface(wrapper)  # we should find a better name
```

**PRO**: No need of a context manager. Avoid indenting long functions.
**CON**: `f` must have the `f(*args, **kwargs)` signature.

Reply:
Great catch thank you :100: I did it less automatically because sometimes you call a transform in the transform and do not execute the QNode. I have added a couple of assert in transforms test.